pkg/providers/aws/aws_provider.go

```diff
@@ -0,0 +1,230 @@
+	defaultRegion       = "us-east-1"
+	defaultDeviceName   = "/dev/xvda"
+	defaultDeviceSize   = 1024  // GiB
+	defaultDeviceType   = "gp2" // General purpose SSD
+)
+
+type InstanceSpec struct {
+	Type  string `json:"type"`
+	AMIID string `json:"amiID"`
+}
+
+type RegionSpec struct {
+	Name            string         `json:"name"`
+	KeyPairName     string         `json:"keyPairName"`
+	SecurityGroupID string         `json:"securityGroupID"`
+	SubnetID        string         `json:"subnetID"`
+	Instances       []InstanceSpec `json:"instances"`
+}
+
+type BlockDeviceSpec struct {
+	DeviceName string `json:"deviceName"` // logica ldevice name e.g. "/dev/xvda"
+	DeviceSize int32  `json:"deviceSize"` // in GiB
+	DeviceType string `json:"deviceType"` // e.g. gp2, gp3, etc.
+}
+
+type MachineSpec struct {
+	Regions []RegionSpec    `json:"regions"`
+	Device  BlockDeviceSpec `json:"device"`
+}
+
+type awsProviderConfig struct {
+	AccessKey       string       `json:"accessKey"`
+	SecretAccessKey string       `json:"secretAccessKey"`
+	UserData        string       `json:"userData,omitempty"`
+	MachineSpec     *MachineSpec `json:"machineSpec"`
+}
+
+type awsProvider struct {
+	config  awsProviderConfig
+	logger  logr.Logger
+	handler AWSHandlerInterface
+}
+
+func AWSProviderFactory(providerInfo string, secretData map[string][]byte, logger logr.Logger) (common.Provider, error) {
+	config := awsProviderConfig{
+		MachineSpec: &MachineSpec{
+			Regions: []RegionSpec{
+				{
+					Name: defaultRegion,
+					Instances: []InstanceSpec{
+						{Type: defualtInstanceType, AMIID: defaultAMI},
+					},
+				},
+			},
+			Device: BlockDeviceSpec{
+				DeviceName: defaultDeviceName,
+				DeviceSize: defaultDeviceSize,
+				DeviceType: defaultDeviceType,
+			},
+		},
+	}
+
+	if configJson, ok := secretData["config"]; ok {
+		if err := json.Unmarshal(configJson, &config); err != nil {
+			return nil, fmt.Errorf("error in provider config json: %w", err)
+		}
+	}
+
+	if config.UserData != "" {
+		// Provided userdata from secret should be base64 encoded
+		decodedBytes, err := base64.StdEncoding.DecodeString(config.UserData)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding userdata: %w", err)
+		}
+
+		config.UserData = string(decodedBytes)
+	}
+
+	awsHandler, err := NewAWSHandler(config.AccessKey, config.SecretAccessKey, logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed creating AWS handler")
+	}
+
+	return &awsProvider{
+		config:  config,
+		logger:  logger,
+		handler: awsHandler,
+	}, nil
+}
+
+func (p *awsProvider) Acquire(poolSize int, poolName string, poolType string) (common.Resource, error) {
+	ctx := context.Background()
+
+	total := 0
+	for _, regionSpec := range p.config.MachineSpec.Regions {
+		count, err := p.handler.GetTotalAmountOfPoolInstancesInRegion(
+			ctx, &PoolFilterParams{Region: regionSpec.Name, PoolName: poolName},
+		)
+		if err != nil {
+			return common.Resource{}, errors.Wrapf(
```

Comment:
any specific reason for not using `fmt.Errorf` (with `%w`)?

Reply:
I think `errors.Wrap` conveys better the purpose of this action and is more convenient, is there a downsize for using it ?

Reply:
Notice that it is currently imported indirectly anyway

Reply:
AFAIK the `errors` package is not a standard library one, and currently [moved to maintance mode](https://github.com/pkg/errors?tab=readme-ov-file#roadmap). For consistency also with the other providers, where fmt.Errorf was used (along with `%w`) for wrapping errors.